library/core/src/alloc/layout.rs

```diff
@@ -172,27 +172,33 @@ impl Layout {
     /// allocate backing structure for `T` (which could be a trait
     /// or other unsized type like a slice).
     ///
+    /// This is an unchecked version of [`for_value`][Self::for_value]
+    /// which takes a raw pointer instead of a reference.
+    ///
     /// # Safety
     ///
-    /// This function is only safe to call if the following conditions hold:
-    ///
-    /// - If `T` is `Sized`, this function is always safe to call.
-    /// - If the unsized tail of `T` is:
-    ///     - a [slice], then the length of the slice tail must be an initialized
-    ///       integer, and the size of the *entire value*
-    ///       (dynamic tail length + statically sized prefix) must fit in `isize`.
-    ///     - a [trait object], then the vtable part of the pointer must point
-    ///       to a valid vtable for the type `T` acquired by an unsizing coercion,
-    ///       and the size of the *entire value*
-    ///       (dynamic tail length + statically sized prefix) must fit in `isize`.
-    ///     - an (unstable) [extern type], then this function is always safe to
-    ///       call, but may panic or otherwise return the wrong value, as the
-    ///       extern type's layout is not known. This is the same behavior as
-    ///       [`Layout::for_value`] on a reference to an extern type tail.
-    ///     - otherwise, it is conservatively not allowed to call this function.
+    /// The provided (possibly wide) pointer must describe a valid value layout.
+    /// Specifically:
+    ///
+    /// - If `T` is a `Sized` type, this function is always safe to call and is
+    ///   equivalent to [`size_of::<T>()`][size_of]. The pointer is unused.
+    /// - If the unsized tail of `T` is a [slice], then the size of the *entire
+    ///   value* (statically sized prefix plus dynamic tail) must fit in `isize`.
+    ///   The pointer may be dangling or null; only the slice length is used.
+    /// - If the unsized tail of `T` is a [trait object], then the wide pointer
+    ///   metadata (the vtable reference) must originate from an unsizing coercion,
```

Comment:
I would also require that the vtable is from an unsizing coercion *for the exact trait given in the pointee type*.

Reply:
That doesn't *quite* work, because dyn upcast pointers should also be accepted. This wording might also exclude those, but I think includes it because `null::<()>() as *const (dyn Destruct + Sync) as *const dyn Destruct` still originates from an unsizing coercion. What this wording doesn't exclude is `transmute<*const dyn Destruct, *const dyn AsRef<str>>(null::<()>())`.

I legitimately don't know how to phrase this to include the former but exclude the latter. I suppose that

> was acquired by unsizing or trait upcasting coercion to this trait object type

*should* suffice, since a trait upcast from an incorrect vtable base is itself UB, but it feels insufficient somehow that I can't quite identify.

Reply:
Ising internal compiler terms the wording would be that the *principal* trait given in the pointee type must match the vtable. I am not sure entirely how that's defined... the trait that's left after removing all marker traits?

Reply:
In the [code](https://github.com/rust-lang/rust/blob/905f5658246ae2f108dee3620ace203c0d5286e1/compiler/rustc_middle/src/ty/predicate.rs#L283-L290) "principal" seems to be just "first trait", but... we accept code where the marker trait comes first
```rust
fn foo(x: &(dyn Send + std::fmt::Debug)) {}
```
and I don't know how that fits together. Cc @lcnr

Reply:
principal is the single non-auto trait of a trait object, if one exists, e.g. even in `dyn Send + Debug` it's still `Debug`.

Reply:
I'm a bit confused that `self[0]` manages to find that, I guess something somewhere is sorting this array to put the non-auto-trait first.

Thanks!

Reply:
yeah, we sort (and iirc deduplicate) the existential predicates of trait objects when lowering their HIR types to the actual `ty` representation